File: 2017/day07.py
import re


def part1(data):
    nodes = parse(data)
    return [n for n in nodes if not any(n in c for _w, c in nodes.values())][0]


def part2(data):
    nodes = parse(data)
    for node in nodes:
        weights = {c: holding(nodes, c) for c in nodes[node][1]}
        # create a children by weight mapping
        freq = {}
        for key, value in weights.items():
            if value not in freq:
                freq[value] = set()
            freq[value].add(key)
        if len(freq) > 1:
            # find the odd child out, and the remaining weight
            # is the expected weight
            (odd,) = min(freq.values(), key=len)
            freq.pop(weights[odd])
            (expect,) = freq.keys()
            # adjust the node's local weight to make it's total match expected
            return nodes[odd][0] + expect - weights[odd]
    return None


def holding(nodes, node):
    weight, children = nodes[node]
    return weight + sum(holding(nodes, n) for n in children)


def parse(data):
    nodes = {}
    for line in data.split("\n"):
        match = re.search(r"(\w+) \((\d+)\)( -> (.*))?", line)
        weight = int(match.group(2))
        children = {c for c in (match.group(4) or "").split(", ") if c}
        nodes[match.group(1)] = (weight, children)
    return nodes

File: 2022/src/day07.rs
use std::collections::HashMap;

#[derive(Debug)]
struct FileSystem {
    paths: Vec<Path>,
}

#[derive(Debug)]
struct Path {
    size: usize,
    parent: usize,
    children: HashMap<String, usize>,
}

pub fn solve() {
    let fs = parse(&std::fs::read_to_string("data/day07.txt").expect("data-file"));

    // part 1
    println!("part 1: {}", part1(&fs));

    // part 2
    println!("part 2: {}", part2(&fs));
}

fn parse(data: &str) -> FileSystem {
    let mut fs = FileSystem {
        paths: vec![Path {
            size: 0,
            parent: 0,
            children: HashMap::new(),
        }],
    };
    let mut cd = 0;
    let mut ls = false;
    for line in data.lines() {
        if line == "$ ls" {
            ls = true;
        } else if line.starts_with("$ cd ") {
            ls = false;
            let (_, name) = line.split_at(5);
            cd = match name {
                "/" => 0,
                ".." => fs.paths[cd].parent,
                _ => *fs.paths[cd].children.get(name).expect("existing-directory"),
            }
        } else if ls {
            let (size, name) = line.split_once(' ').expect("ls-output-space");
            // ignore any paths we have already ls'ed
            if !fs.paths[cd].children.contains_key(name) {
                // insert the new node
                let child = fs.paths.len();
                fs.paths.push(Path {
                    size: if size != "dir" {
                        size.parse().expect("file-size")
                    } else {
                        0
                    },
                    parent: cd,
                    children: HashMap::new(),
                });
                fs.paths[cd].children.insert(String::from(name), child);
            }
        } else {
            panic!("unknown-line");
        }
    }
    fs
}

fn part1(fs: &FileSystem) -> usize {
    let mut cache = vec![0; fs.paths.len()];
    (0..fs.paths.len())
        .filter(|&id| fs.paths[id].size == 0)
        .map(|id| tally(fs, &mut cache, id))
        .filter(|&s| s < 100000)
        .sum()
}

fn part2(fs: &FileSystem) -> usize {
    let mut cache = vec![0; fs.paths.len()];
    let total = tally(fs, &mut cache, 0);
    (0..fs.paths.len())
        .filter(|&id| fs.paths[id].size == 0)
        .map(|id| tally(fs, &mut cache, id))
        .filter(|&s| 70000000 - (total - s) >= 30000000)
        .min()
        .expect("large-enough-directory")
}

fn tally(fs: &FileSystem, cache: &mut Vec<usize>, id: usize) -> usize {
    let child = &fs.paths[id];
    if child.size != 0 {
        child.size
    } else {
        if cache[id] == 0 {
            cache[id] = child
                .children
                .values()
                .map(|&id| tally(fs, cache, id))
                .sum();
        }
        cache[id]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const DATA: &str = "\
$ cd /
$ ls
dir a
14848514 b.txt
8504156 c.dat
dir d
$ cd a
$ ls
dir e
29116 f
2557 g
62596 h.lst
$ cd e
$ ls
584 i
$ cd ..
$ cd ..
$ cd d
$ ls
4060174 j
8033020 d.log
5626152 d.ext
7214296 k
";

    lazy_static! {
        static ref FS: FileSystem = parse(DATA);
    }

    #[test]
    fn test_part1() {
        assert_eq!(part1(&FS), 95437);
    }

    #[test]
    fn test_part2() {
        assert_eq!(part2(&FS), 24933642);
    }
}
